lib/Service/ImageService.php

```diff
@@ -0,0 +1,676 @@
+		try {
+			$share = $this->shareManager->getShareByToken($shareToken);
+			return ($share->getShareType() === IShare::TYPE_LINK && $share->getPermissions() & Constants::PERMISSION_UPDATE);
+		} catch (ShareNotFound $e) {
+			return false;
+		}
+	}
+
+	/**
+	 * Download an image from a link and place it in a given folder
+	 * @param Folder $saveDir
+	 * @param string $link
+	 * @param File $textFile
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 */
+	private function downloadLink(Folder $saveDir, string $link, File $textFile): array {
+		$fileName = (string) time();
+		$savedFile = $saveDir->newFile($fileName);
+		$resource = $savedFile->fopen('wb');
+		$res = $this->simpleDownload($link, $resource);
+		if (is_resource($resource)) {
+			fclose($resource);
+		}
+		$savedFile->touch();
+		if (isset($res['Content-Type'])) {
+			if (in_array($res['Content-Type'], ImageController::IMAGE_MIME_TYPES)) {
+				if ($res['Content-Type'] === 'image/jpeg') {
+					$fileName = $fileName . '.jpg';
+				} elseif ($res['Content-Type'] === 'image/x-xbitmap' || $res['Content-Type'] === 'image/x-ms-bmp') {
+					$fileName = $fileName . '.bmp';
+				} elseif ($res['Content-Type'] === 'image/svg+xml') {
+					$fileName = $fileName . '.svg';
+				} else {
+					$ext = preg_replace('/^image\//i', '', $res['Content-Type']);
+					$fileName = $fileName . '.' . $ext;
+				}
+				$targetPath = $saveDir->getPath() . '/' . $fileName;
+				$savedFile->move($targetPath);
+				// get file type and name
+				return [
+					'name' => $fileName,
+					'id' => $savedFile->getId(),
+					'documentId' => $textFile->getId(),
+				];
+			} else {
+				$savedFile->delete();
+				return [
+					'error' => 'Unsupported file type',
+				];
+			}
+		} elseif (isset($res['error'])) {
+			$savedFile->delete();
+			return $res;
+		} else {
+			$savedFile->delete();
+			return [
+				'error' => 'Link download error',
+			];
+		}
+	}
+
+	/**
+	 * Get or create the user-specific attachment folder
+	 *
+	 * @param string $userId
+	 * @return Folder|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getOrCreateTextDirectory(string $userId): ?Folder {
+		$userFolder = $this->rootFolder->getUserFolder($userId);
+		if ($userFolder->nodeExists('/Text')) {
+			$node = $userFolder->get('Text');
+			if ($node instanceof Folder) {
+				return $node;
+			} else {
+				return null;
+			}
+		} else {
+			return $userFolder->newFolder('/Text');
+		}
+	}
+
+	/**
+	 * Get or create file-specific attachment folder
+	 *
+	 * @param File $textFile
+	 * @return Folder|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getOrCreateAttachmentDirectoryForFile(File $textFile): ?Folder {
```

Comment:
Since `getOrCreateAttachmentDirectoryForFile()` and `getAttachmentDirectoryForFile()` share a lot of code, wouldn't it be better to merge them and instead e.g. add a boolean parameter `$create`?

Reply:
:+1: